### PR TITLE
feat: Implement route index management and rebuilding functionality

### LINKED ISSUE #128: Phase 2: Build Index on Route Create/Update
## Problem

Index table exists but is empty. Need logic to populate it when routes are created/updated.

## Proposed Solution

Create `RouteIndexService` with index building logic. Integrate with `RouteService` to trigger builds. Add admin endpoint and Celery task for manual/async building.

## Service Logic

**New file: `backend/app/services/route_index_service.py`**

**Core method:**
```python
async def build_route_station_index(self, route_id: UUID) -> None:
    """
    Build inverted index for a route by expanding segments to all intermediate stations.
    
    Algorithm:
    1. Delete existing index entries for this route (in transaction)
    2. For each segment pair (current → next):
       - Get Line.routes JSON for current.line_id
       - Find ALL orderedLineRoutes containing BOTH stations
       - For EACH matching route variant:
         - Extract all stations between current and next (inclusive)
         - Insert index entry for each (line_tfl_id, station_naptan)
    3. Commit transaction
    
    Note: If multiple route variants exist (e.g., Northern Line Bank/Charing Cross branches),
    we store ALL stations from ALL variants. This means alerts trigger if ANY branch is disrupted.
    
    Complexity: O(segments × route_variants × stations_per_variant)
    Expected time: 50-100ms per route
    """
```

**Integration points:**
- `RouteService.create_route()` → trigger build
- `RouteService.upsert_segments()` → trigger rebuild
- `RouteService.delete_route()` → cascade delete handles cleanup

## API Changes

**New admin endpoint:**
```
POST /admin/routes/rebuild-index
Body: {"route_id": "uuid"} (optional - if null, rebuild ALL routes)
Response: {"rebuilt_count": 10, "failed_count": 0}
```

## Celery Task

**File: `backend/app/celery/tasks.py`**
```python
@celery_app.task(bind=True, max_retries=3)
def build_route_index_task(self, route_id: str) -> dict:
    """Async wrapper for index building."""
    return asyncio.run(_build_route_index_async(UUID(route_id)))
```

## Example Algorithm (Pseudocode)

```python
async def build_route_station_index(self, route_id: UUID) -> None:
    route = await self.get_route_with_segments(route_id)
    
    # Clear existing index entries
    await self.db.execute(
        delete(RouteStationIndex).where(RouteStationIndex.route_id == route.id)
    )
    
    for i in range(len(route.segments) - 1):
        current = route.segments[i]
        next_seg = route.segments[i + 1]
        
        if not current.line_id:
            continue  # Skip destination-only segments
        
        # Get line with routes data
        line = await self.db.get(Line, current.line_id)
        if not line.routes:
            continue
        
        # Find ALL matching variants (may be multiple for branched lines)
        for variant in line.routes["routes"]:
            stations = variant["stations"]  # Ordered list of NaPTAN codes
            
            # Check if this variant contains both stations
            try:
                start_idx = stations.index(current.station.tfl_id)
                end_idx = stations.index(next_seg.station.tfl_id)
            except ValueError:
                continue  # This variant doesn't contain both stations
            
            # Insert index entries for ALL stations between start and end
            start, end = min(start_idx, end_idx), max(start_idx, end_idx)
            for station_naptan in stations[start:end+1]:
                self.db.add(RouteStationIndex(
                    route_id=route.id,
                    line_tfl_id=line.tfl_id,
                    station_naptan=station_naptan,
                    line_data_version=line.last_updated,
                ))
    
    await self.db.commit()
```

## Multi-Branch Handling

**Example:** Morden → Camden Town (Northern Line)
- Northern has 2 branches: Bank and Charing Cross
- Both contain Morden and Camden Town
- **We store stations from BOTH branches**
- If Bank branch disrupted, user IS alerted (even if Charing Cross fine)
- Rationale: User didn't specify branch, might use either

## Task List

- [ ] Create `RouteIndexService` with `build_route_station_index()` method
- [ ] Implement station expansion logic using Line.routes JSON
- [ ] Handle multiple route variants (store ALL matching variants' stations)
- [ ] Handle hub station resolution (use existing helpers)
- [ ] Handle segments with line_id=NULL (skip, destination only)
- [ ] Integrate with `RouteService.create_route()`
- [ ] Integrate with `RouteService.upsert_segments()`
- [ ] Add admin API endpoint `POST /admin/routes/rebuild-index`
- [ ] Add Celery task `build_route_index_task`
- [ ] Write unit tests for index building (mock Line.routes data)
- [ ] Write unit tests for multi-branch handling (Northern Line example)
- [ ] Write integration tests for route create → index built
- [ ] Write integration tests for route update → index rebuilt
- [ ] Write tests for edge cases (missing Line.routes, no matching variant, NULL line_id)
- [ ] Write performance test (build index for 20-station route < 100ms)
- [ ] Verify 100% code coverage for new service

## ADR Updates

**File: `docs/adr/08-background-jobs.md`**
- Add section: **Async Route Index Building**
- Context: Index building (100-500ms) too slow for sync API responses
- Decision: Use Celery task for async building, triggered on route changes
- Consequences: Fast API, automatic rebuild, small window before new route indexed

## Testing Requirements (100% Coverage)

- Unit: Index building with various Line.routes structures (10+ test cases)
- Unit: Multi-branch handling (Northern Line Bank/Charing Cross example)
- Unit: Single branch lines (Victoria, Central)
- Unit: Hub station handling
- Unit: NULL line_id handling (skip segment)
- Integration: Create route → verify index entries correct
- Integration: Create route on branched line → verify ALL branch stations indexed
- Integration: Update segments → verify index rebuilt
- Integration: Delete route → verify index entries gone
- Integration: Admin endpoint rebuild (single + all routes)
- Performance: Build time < 100ms for typical route
- Edge cases: Missing Line.routes, no matching variant, circular routes

## Dependencies

Requires Phase 1 to be completed first.

## Related Issues

Part of #125 (Route Station Inverted Index)

## Estimated Time

3-4 days

### PR DESCRIPTION
**Pull request by mnbf9rca**
- Added a new endpoint to the admin API for rebuilding route station indexes, allowing for single or all routes to be rebuilt.
- Introduced RouteIndexService to handle the logic for building and maintaining route station indexes.
- Updated Celery tasks to include a task for rebuilding route indexes, which can be triggered manually or scheduled.
- Created new schemas for the response of the rebuild operation.
- Enhanced the RouteService to automatically rebuild indexes after route modifications.
- Added comprehensive tests for the RouteIndexService, covering various scenarios including multi-variant lines and error handling.

Closes #128

## Summary by Sourcery

Implement end-to-end route station index management by introducing a dedicated service, API endpoint, and Celery task for rebuilding inverted (line, station)→route indexes, integrating automatic index updates on route changes, and adding full test coverage.

New Features:
- Add admin API endpoint POST /admin/routes/rebuild-indexes to rebuild station indexes for single or all routes
- Introduce RouteIndexService to build and maintain route station inverted indexes
- Implement Celery task rebuild_route_indexes_task for scheduled or manual index rebuilding

Enhancements:
- Enhance RouteService to automatically rebuild indexes after segment upserts, updates, and deletions
- Add type-safe task result protocols and TypedDicts for structured logging and retry behavior
- Improve session management and logging in async task implementations

Tests:
- Comprehensive tests for RouteIndexService pure functions, index building across varied line topologies, rebuild_routes logic, and auto-commit/error scenarios
- Add tests for Celery rebuild task behavior, including success, partial failure, retries, and session cleanup
- Add unit and integration tests for the admin rebuild-indexes endpoint, covering permission, error handling, and full database workflows

### CODE REVIEW

File: backend/app/services/route_index_service.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        >>> deduplicate_preserving_order([])
+        []
+    """
+    seen: set[str] = set()
+    result: list[str] = []
+    for item in items:
+        if item not in seen:
+            seen.add(item)
+            result.append(item)
+    return result
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/route_index_service.py:494]: ✅ Fixed. Replaced the inline deduplication logic (lines 367-369) with a call to `deduplicate_preserving_order()`. The method now collects all stations using `.extend()` and then calls the pure function once after the loop. This eliminates code duplication and follows the pure function architecture guideline.

File: backend/app/api/admin.py [hunk 2]
```diff
@@ -515,3 +518,73 @@ async def get_engagement_metrics(
...
+                    failed_count += 1
+                    errors.append(f"Route {route.id}: {exc!s}")
+
+        return RebuildIndexesResponse(
+            success=failed_count == 0,
+            rebuilt_count=rebuilt_count,
+            failed_count=failed_count,
+            errors=errors,
+        )
+
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/api/admin.py]: ✅ Fixed. Extracted shared `rebuild_routes()` method in RouteIndexService that both the admin endpoint and Celery task now use. This eliminates ~21 lines of duplication and ensures consistent behavior. The method handles error tracking and statistics collection in one place, making future maintenance easier.

File: backend/app/api/admin.py [hunk 3]
```diff
@@ -515,3 +518,73 @@ async def get_engagement_metrics(
...
+    failed_count = 0
+    errors: list[str] = []
+
+    try:
+        if route_id:
+            # Rebuild single route
+            try:
+                await index_service.build_route_station_index(route_id, auto_commit=True)
+                rebuilt_count = 1
+            except ValueError as exc:
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/api/admin.py]: ✅ Fixed. The shared `rebuild_routes()` method now uses consistent `Exception` handling for both single and bulk rebuild operations. This ensures database errors and other exceptions are properly tracked in `failed_count` and `errors` rather than propagating as HTTP 500 errors, providing better UX.

File: backend/app/services/route_index_service.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+                "build_route_station_index_completed",
+                route_id=str(route_id),
+                entries_created=entries_created,
+                segments_count=len(route.segments),
+                auto_commit=auto_commit,
+            )
+
+            return {
+                "entries_created": entries_created,
+                "segments_processed": len(route.segments) - 1,  # Pairs count
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/route_index_service.py]: ✅ Fixed. Changed `_process_segments()` to return `tuple[int, int]` (entries_created, pairs_processed) and added a counter that increments only when a pair is actually processed. The method now accurately tracks processed pairs instead of returning the theoretical maximum `len(segments) - 1`. This fixes the bug where pairs starting from segments with `line_id=None` were incorrectly counted.

File: backend/tests/test_route_index_service.py [hunk 5]
```diff
@@ -0,0 +1,639 @@
...
+        ]
+        db_session.add_all(segments)
+        await db_session.commit()
+
+        # Build index
+        service = RouteIndexService(db_session)
+        result = await service.build_route_station_index(route.id)
+
+        # Verify result statistics
+        assert result["segments_processed"] == 3  # 4 segments = 3 pairs
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/test_route_index_service.py]: ✅ Fixed. Updated test assertion to `assert result["segments_processed"] == 2` with a comment explaining why 4 segments results in 2 processed pairs (pair starting from segment 2 is skipped because segment 2 has `line_id=None`). The test now validates the correct behavior after fixing the service bug.

**Comment by mnbf9rca** [comment]
## PR Review Comments - All Addressed ✅

All 5 review comments have been addressed with the following fixes:

### 1. Unused Helper Functions (Comment #1)
**Fixed:** Replaced inline deduplication logic with `deduplicate_preserving_order()` pure function
- Changed from manual loop checking `if station not in all_stations` to using the exported helper
- Follows pure function architecture guideline
- Reduces code duplication

### 2. Code Duplication (Comment #2)
**Fixed:** Extracted shared `rebuild_routes()` method in `RouteIndexService`
- Eliminates ~21 lines of duplicated code between admin endpoint and Celery task
- Both now call `index_service.rebuild_routes()` for consistent behavior
- Improves maintainability

### 3. Inconsistent Exception Handling (Comment #3)
**Fixed:** Consistent `Exception` handling in shared `rebuild_routes()` method
- Both single and bulk rebuilds now catch all `Exception` types (not just `ValueError`)
- Errors properly tracked in `failed_count` and `errors` response
- Better UX - avoids HTTP 500 for recoverable errors

### 4. Misleading `segments_processed` Count (Comment #4)
**Fixed:** Accurate tracking of processed segment pairs
- Changed `_process_segments()` to return `tuple[int, int]` 
- Added `pairs_processed` counter that increments only when pair is actually processed
- No longer returns theoretical maximum `len(segments) - 1`
- Correctly handles segments with `line_id=None`

### 5. Incorrect Test Assertion (Comment #5)
**Fixed:** Updated test to expect correct behavior
- Changed assertion from `segments_processed == 3` to `segments_processed == 2`
- Added explanatory comment about why 4 segments = 2 processed pairs
- Test now validates correct service behavior

## Test Results ✅
- All 32 tests passing (19 route_index_service + 13 celery_tasks)
- Pre-commit hooks passing (ruff, mypy, formatting)
- No regressions introduced

## Code Quality Improvements
- ✅ Eliminated code duplication (DRY principle)
- ✅ Used pure functions for testability
- ✅ Consistent error handling
- ✅ Accurate metrics tracking
- ✅ Type-safe with mypy --strict
- ✅ All tests validating correct behavior

**Comment by mnbf9rca** [comment]
## Update: Comprehensive Tests Added for rebuild_routes ✅

Added 6 new tests specifically for the rebuild_routes method to ensure proper test coverage after the refactoring:

### New Test Coverage
1. test_rebuild_single_route_success - Verifies single route rebuild works correctly
2. test_rebuild_single_route_not_found - Ensures errors are properly tracked when route doesn't exist
3. test_rebuild_all_routes_success - Tests bulk rebuild of all routes
4. test_rebuild_all_routes_partial_failure - Validates error handling when some routes fail
5. test_rebuild_all_routes_empty - Handles edge case of no routes
6. test_rebuild_routes_exception_handling - Confirms consistent exception handling

### Test Results
- All 25 route_index_service tests passing - 19 original + 6 new
- Pre-commit hooks passing
- Service method coverage: 80.24 percent improved from baseline
- Tests validate statistics tracking, single vs bulk rebuild behavior, error propagation, and edge cases

The refactored rebuild_routes method now has dedicated test coverage, and both the admin endpoint and Celery task benefit from this shared, well-tested logic.

**Comment by mnbf9rca** [comment]
## Final Update: Added Admin Endpoint Integration Tests ✅

Added 5 integration tests for the admin rebuild indexes endpoint to complement the service layer tests.

### Complete Test Coverage Summary

**Service Layer Tests** - test_route_index_service.py (25 tests):
- 19 original tests for build_route_station_index
- 6 new tests for rebuild_routes method

**Integration Tests** - test_admin_route_index.py (5 NEW tests):
1. test_rebuild_indexes_success_single_route - Verify endpoint calls service correctly for single route
2. test_rebuild_indexes_success_all_routes - Verify bulk rebuild via endpoint
3. test_rebuild_indexes_partial_failure - Error response formatting
4. test_rebuild_indexes_requires_admin - Authorization check
5. test_rebuild_indexes_exception_handling - Exception to HTTP 500 conversion

### Test Results
- All 30 tests passing (25 service + 5 endpoint)
- route_index_service.py coverage: 93.41 percent
- Pre-commit hooks passing on all files
- Admin endpoint properly tested as thin wrapper around service

The admin endpoint and Celery task are now lightweight wrappers around the well-tested rebuild_routes service method, following proper separation of concerns.

File: backend/app/services/route_index_service.py [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+            )
+            return 0, 0
+
+        entries_created = 0
+        pairs_processed = 0
+
+        # Process each consecutive segment pair
+        for i in range(len(segments) - 1):
+            current_segment = segments[i]
+            next_segment = segments[i + 1]
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/route_index_service.py:267]: ✅ Fixed. Added defensive null checks for both current_segment.station and next_segment.station before accessing their tfl_id attributes (lines 274-284). The code now skips segment pairs with missing station data and logs a warning with detailed information about which segments are affected.

File: backend/app/api/admin.py [hunk 7]
```diff
@@ -515,3 +517,51 @@ async def get_engagement_metrics(
...
+    Args:
+        route_id: Optional route UUID. If provided, rebuilds only that route. If None, rebuilds all routes.
+        admin_user: Authenticated admin user
+        db: Database session
+
+    Returns:
+        Statistics about the rebuild operation
+
+    Raises:
+        HTTPException: 403 if not admin, 404 if route_id not found
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/api/admin.py]: ✅ Fixed. Updated docstring to accurately reflect that individual route failures are reported in the errors list rather than raising 404 exceptions. Also improved exception handling to re-raise HTTPExceptions (preserving their status codes like 403) while catching only truly unexpected errors like database connection failures.

**Comment by mnbf9rca** [comment]
## Code Review Comments Addressed

Thank you for the detailed review! I've addressed all the feedback:

### Overall Comments

**1. Celery task wrapper using asyncio.run:**
- **Resolution**: Added documentation explaining this is the standard pattern for Celery tasks calling async code
- Each task runs in its own worker thread, so asyncio.run() creating a new event loop per task is not blocking in the worker context
- See updated comments in `backend/app/celery/tasks.py:96-98` and lines 181-183

**2. Admin endpoint exception handling:**
- **Resolution**: Improved exception handling to re-raise HTTPExceptions while catching only truly unexpected errors
- Now preserves correct status codes (403, 404, etc.) by re-raising HTTPExceptions
- Only converts unexpected exceptions (e.g., database connection loss) to HTTP 500
- Service-level errors are returned in the result dict, not raised as exceptions
- See `backend/app/api/admin.py:565-574`

**3. Synchronous route index rebuilding:**
- **Resolution**: Added ADR documenting the hybrid approach (sync for CRUD, async for bulk)
- The design is intentional: synchronous rebuilding during route changes ensures transactional safety (index + segments are atomic)
- Background Celery task is used for bulk admin rebuilds to avoid blocking API
- Rebuild time is only 50-100ms per route, acceptable for API responses
- See new ADR section in `docs/adr/08-background-jobs.md`

### Individual Comments

**Null check for segment.station:**
- Added defensive null checks for both current_segment.station and next_segment.station
- Logs warnings with detailed info if station data is missing
- See `backend/app/services/route_index_service.py:274-284`

### Test Results
- ✅ All 903 tests passing
- ✅ Overall coverage: 98.10% (exceeds 95% requirement)
- ✅ All pre-commit hooks passing (mypy, ruff, prettier, etc.)

File: backend/app/services/route_index_service.py [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+                from_idx = stations.index(from_station_naptan)
+                to_idx = stations.index(to_station_naptan)
+            except ValueError:
+                # This variant doesn't contain both stations, try next variant
+                continue
+
+            # Extract all stations between from and to (inclusive)
+            start_idx = min(from_idx, to_idx)
+            end_idx = max(from_idx, to_idx)
+            variant_stations = stations[start_idx : end_idx + 1]
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/route_index_service.py:393]: Thanks for the feedback. I've considered this suggestion but decided to keep the inline implementation for the following reasons:

**Different Error Handling Patterns:**
The `_expand_segment_to_stations` method needs to search through multiple route variants and collect results from ALL matching variants. It can't use `find_stations_between()` as a simple replacement because:
- `find_stations_between()` returns None if either station isn't found, designed for single-variant searches
- This method must try multiple variants and continue searching if a variant doesn't contain both stations
- The logic flow is "try variant → extract stations → continue to next variant", not "try once → return"

**KISS Principle:**
Following the hobby project guidelines, the inline 3-line extraction is actually clearer in the method's context. Refactoring to use the pure function would add indirection without clear benefit.

**Not Actually Used in Production:**
The pure function `find_stations_between()` is only used in tests, never in production code, suggesting it was created for testing/future use rather than as a shared utility for this specific case.

**Solution:**
I've added a comment (lines 383-386) explaining why the pure function isn't used here. This documents the reasoning for future maintainers.

File: backend/app/services/route_index_service.py [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+
+        Returns:
+            Route instance or None if not found
+        """
+        result = await self.db.execute(
+            select(Route)
+            .where(Route.id == route_id)
+            .options(
+                selectinload(Route.segments).selectinload(RouteSegment.station),
+                selectinload(Route.segments).selectinload(RouteSegment.line),
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/route_index_service.py:222]: Excellent catch! This is a valid concern - the ordering dependency is critical but not obvious from the query code.

**Why This Matters:**
The `_process_segments` method (lines 233-335) absolutely depends on segments being ordered by sequence. It uses `segments[i]` and `segments[i + 1]` to process adjacent pairs. If segments were unordered, this would create incorrect index entries linking non-adjacent stations.

**Solution:**
I've added a clarifying comment (lines 213-216) documenting:
- That segments MUST be ordered by sequence
- Where the ordering comes from (Route.segments relationship in route.py:68)
- Why this is critical (would create wrong index entries if broken)

**Why Comment Over Query Refactor:**
Following the KISS principle for this hobby project, a clarifying comment is safer than refactoring the query with explicit `.order_by()`. The suggested syntax would require careful testing with SQLAlchemy eager loading to ensure it doesn't change query behavior.

The comment makes the implicit dependency explicit for future maintainers.

**Comment by mnbf9rca** [comment]
## Unresolved PR Comments Addressed ✅

Addressed the remaining 2 unresolved nitpick comments from code review:

### Comment 1: Line 373 - Duplicate Logic in `_expand_segment_to_stations`
**Reviewer's Concern:** The method reimplements station extraction logic that exists in the `find_stations_between` pure function.

**Resolution:** REJECTED with explanatory comment
- Added clarifying comment (lines 383-386) explaining why the pure function isn't used here
- Different error handling patterns: method needs to search multiple route variants and collect ALL matches
- The inline 3-line implementation is clearer in the multi-variant search context
- Following KISS principle for this hobby project
- Replied to review comment with detailed reasoning

### Comment 2: Line 218 - Implicit Ordering Dependency
**Reviewer's Concern:** Query relies on relationship-level `order_by` from Route model, which isn't visible in the query code.

**Resolution:** ACCEPTED with clarifying comment
- Added comment (lines 213-216) documenting the critical ordering dependency
- Explains that `_process_segments` depends on segments being ordered by sequence
- Documents where ordering comes from (Route.segments relationship in route.py:68)
- Notes that incorrect ordering would create wrong index entries
- Marked review thread as resolved

### Test Results ✅
- All 29 route_index_service tests passing
- Pre-commit hooks passing (ruff, mypy, formatting)
- No functional changes, only documentation improvements
- Ready for review and merge
